### ISSUE TITLE
First Look: Understanding DRF OpenID Connect Auth

### ISSUE THREAD

**Issue by mbrahimi617** [ByteInternet/drf-oidc-auth#36]
I have been looking a few different projects for OpenID implementation until I stumbled across this one, and decided I should check out the source since there isn't a whole lot of documentation to know if it will suit my needs. After reading, a few questions came to mind:

1. Where do I specify a client secret to secure the connection?
2. Where are the actual API views and urls for a login and callback?
3. How do I best redirect from these views to my frontend upon logging in?

**Comment by kerrermanisNL** [comment]
Hey @mbrahimi617 

1. You don't have to specify a client secret. The DRF system should use the token you've provided during authentication there (so provide something like `headers={'Authorization': 'JWT myjwt'}`). This will then use the authentication (`BearerTokenAuthentication` or `JSONWebTokenAuthentication` for example) method you've chosen to retrieve the correct information from the OP (using the JWT/Bearer token you've provided in the `Authorization` header).

2. I'm not sure if I understand you correctly, but since this is mainly meant for web API purposes there are no actual login views. A user has to retrieve his bearer/jwt from the authentication provider (OP) and then use this value in the Authorization header in a request to the API `headers={'Authorization': 'JWT myjwt'}` or `headers={'Authorization': 'Bearer mybearer'}`. The API will then retrieve the correct user information from the OP using this token.

3. Again, either I don't understand or it's not the way you would use this library. Since this is meant for API's there should generally not be an authentication view. The user should already be authenticated when making requests to the API (i.e. providing the authorization headers).

If you want to use this library as a "people can go to my api url and if they're not logged in they can use some OIDC button to continue" I don't think that's what the library is for :) Perhaps a better way to explain it is that this module adds two authentication classes for DRF, namely: `oidc_auth.authentication.JSONWebTokenAuthentication` and `oidc_auth.authentication.BearerTokenAuthentication`, but it provides no actual implementation of something using these authentication methods (i.e. a login view or anything like that). It's just one of the authentication methods that's tried when you need to be authenticated for a certain URL. It will then, based on the request, try a few things (i.e. regular django token/session authentication and oidc authentication). If it cannot authenticate a user for this request then it will just not authenticate the user.

Hope that makes sense, if you have any further questions feel free to ask.

**Comment by mbrahimi617** [comment]
Does this package not associate a successfully authentication instance with a Django User model? Ideally I was looking for an OpenID Connect client such that I could use SSO provided by Ping Identity with a discovery document but also use Django Users for permissions on my own local API? To be honest, I have no need for the ID token or access token after logging in because there is no external API.

**Comment by kerrermanisNL** [comment]
Yes, it does associate a successful authentication with a Django user instance, but it will not _create_ a Django user for this if it is missing. I.e. the user already needs to exists in the API system before you can authenticate the user. I'm not too familiar with Ping identity but if they send some user creation events to a webhook you could set something up there. Once you have the user in DRF you can use the permission system etc. as you usually would.

So in short:
1. Make sure the user exists in the system (i.e. let Ping Identity send an event to a webhook, or let some other system/OP sync this information with your API).
2. Send requests to the API from your systems using the logged in user's Bearer/JWT token (depending on what you're using). 

Hope that helps?
